test/abstract_syntax_tree_location_data.coffee

```diff
@@ -3349,3 +3349,178 @@ test "AST location data as expected for Switch node", ->
               column: 24
       ]
     ]
+test "AST location data as expected for Code node", ->
+  testAstLocationData '''
+    (a) ->
+      b
+      c()
+  ''',
+    type: 'FunctionExpression'
+    params: [
+      start: 1
+      end: 2
+      range: [1, 2]
+      loc:
+        start:
+          line: 1
+          column: 1
+        end:
+          line: 1
+          column: 2
+    ]
+    body:
+      body: [
+        start: 9
+        end: 10
+        range: [9, 10]
+        loc:
+          start:
+            line: 2
+            column: 2
+          end:
+            line: 2
+            column: 3
+      ,
+        start: 13
+        end: 16
+        range: [13, 16]
+        loc:
+          start:
+            line: 3
+            column: 2
+          end:
+            line: 3
+            column: 5
+      ]
+      start: 7
+      end: 16
+      range: [7, 16]
+      loc:
+        start:
+          line: 2
+          column: 0
+        end:
+          line: 3
+          column: 5
+
+  testAstLocationData '''
+    -> a
+  ''',
+    type: 'FunctionExpression'
+    body:
+      start: 2
+      end: 4
+      range: [2, 4]
+      loc:
+        start:
+          line: 1
+          column: 2
+        end:
+          line: 1
+          column: 4
+    start: 0
+    end: 4
+    range: [0, 4]
+    loc:
+      start:
+        line: 1
+        column: 0
+      end:
+        line: 1
+        column: 4
+
+  testAstLocationData '''
+    (
+      a,
+      [
+        b
+        c
+      ]
+    ) ->
+      d
+  ''',
+    type: 'FunctionExpression'
+    params: [
+      start: 4
+      end: 5
+      range: [4, 5]
+      loc:
+        start:
+          line: 2
+          column: 2
+        end:
+          line: 2
+          column: 3
+    ,
+      elements: [
+        start: 15
+        end: 16
+        range: [15, 16]
+        loc:
+          start:
+            line: 4
+            column: 4
+          end:
+            line: 4
+            column: 5
+      ,
+        start: 21
+        end: 22
+        range: [21, 22]
+        loc:
+          start:
+            line: 5
+            column: 4
+          end:
+            line: 5
+            column: 5
+      ]
+      start: 9
+      end: 26
+      range: [9, 26]
+      loc:
+        start:
+          line: 3
+          column: 2
+        end:
+          line: 6
+          column: 3
+    ]
+    start: 0
+    end: 35
+    range: [0, 35]
+    loc:
+      start:
+        line: 1
+        column: 0
+      end:
+        line: 8
+        column: 3
+
+  testAstLocationData '''
+    ->
+  ''',
+    type: 'FunctionExpression'
+    body:
+      start: 2
+      end: 2
+      range: [2, 2]
+      loc:
+        start:
+          line: 1
+          column: 2
+        end:
+          line: 1
+          # column: 2 TODO: make this accurate/should match range?
```

Comment:
The end column location data is off-by-one for the function body block since its end location is demarcated by a generated zero-length token:

Basically, its location data would look something like:
```
range: [2, 2]
first_line: 0
first_column: 2
last_line: 0
last_column: 2
```

But then since Jison line/column data is inclusive (while Babel's is exclusive), we do "column + 1" when converting to AST location data. For a zero-length token, that causes wonkiness/disagreement between `range` and the line/column info in the AST location data, eg here you'll end up with:
```
range: [2, 2]
loc:
  start:
    line: 1
    column: 2
  end:
    line: 1
    column: 3
```

I'd started trying to combat the "column + 1" hack by giving tokens an `endsLine` property to know whether "column + 1" is correct (as opposed to "line + 1, column 0" for a token that ends a line (including the actual newline)) - that's on my [ast-block-location-data](https://github.com/helixbass/copheescript/commits/ast-block-location-data) branch if you're curious

However, this case makes me think that approach won't work/be enough, since it wouldn't do anything about this zero-width token case. So then the other approach I'd had in mind was just to go ahead and add additional exclusive end line/column properties to token location data (eg `last_line_exclusive`/`last_column_exclusive`) - from my understanding we can't just replace the existing `last_line`/`last_column` (or convert them to be exclusive) since Jison expects them to be present and to be inclusive

So maybe I'll explore that approach in a different branch. Tracking the exclusive last line/column of tokens would in theory get rid of both the potential inaccuracy of "column + 1" for line-ending tokens (that we've already discussed in previous PRs) as well as this zero-length generated token end line/column location data issue

Reply:
Sounds good. Could you perhaps track in `grammar.coffee` when a zero-width token is passed in? Or maybe in the lexer, see if a token is followed by a zero-width token and preserve that info through Jison? Just trying to think of other solutions.

Reply:
Ya just always explicitly tracking the exclusive last line/column location data is the conceptually simplest approach I can see. I started implementing that and wasn't too bad

The issue (that I was afraid I might run into) that I found is that Jison doesn't just blindly pass along the entire location data object, it explicitly "forwards" `first_line`/`first_column`/`last_line`/`last_column` (and optionally `range`). So Jison is thwarting our attempts to pass through `last_line_exclusive`/`last_column_exclusive` every time `LOC()` gets called

Only a minimal change (two lines if you hard-code `last_line_exclusive`/`last_column_exclusive` rather than eg adding a new option to specify them) to Jison is required to pass those through as well - on my [exclusive-end-line-column](https://github.com/helixbass/copheescript/compare/code-ast...exclusive-end-line-column) branch, I have all the tests passing (including updating these AST location data tests to have the correct end column) against a version of the parser compiled against a temporarily patched version of Jison

So... what is the best way to proceed? Jison looks like it's not actively maintained. So submitting a PR to add an option allowing for passing through additional location data keys doesn't seem like it would get merged. Would there be a downside to forking it and using the forked version as our dependency?

Otherwise I'm not sure if there's a more devious way to smuggle the `last_line_exclusive`/`last_column_exclusive` through the parser

Reply:
Why not smuggle them through like you do the other data, as properties on the string object? Or you can use `addDataToNode`, the method I’m using to pass through the comments: https://github.com/jashkenas/coffeescript/blob/master/src/helpers.coffee#L135-L153

Reply:
I'm not seeing any visibility of the complete token location data objects in `addDataToNode()` - the `first`/`last` args are the "filtered" versions of the location data that don't include `last_line_exclusive`/`last_column_exclusive`. And I didn't see anything relevant sniffing around on the `parserState` arg

So let's think about if we could use that typical smuggling technique. We could modify the call to `addDataToNode()` in the grammar to also include args for eg `$1` (the token value) in addition to `@$1` (the filtered location data that is the current `first`/`last` args)

At that point I believe every single token would need to use the smuggling-via-`String()`-object technique, since any token could get used in a `LOC()` call for its location data

So the question I think is whether it's feasible to use that technique *everywhere*. If that would mean needing to manually call `.toString()` on every single grammar rule, that seems ugly enough that I'd question what the downside is of forking Jison to avoid having to do that

Although here's a thought - our grammar `o()` function is already doing some magic of rewriting the actual function body. Maybe we could generate the `.toString()` calls on any "bare" references to eg `$1`...?